### PR TITLE
scipy/stats: Add laplace_asymmetric continuous distribution

### PR DESCRIPTION
**Pull request by essandess**
scipy/stats: Add laplace_asymmetric continuous distribution

#### What does this implement/fix?

This PR adds the [asymmetric Laplace distribution](https://en.wikipedia.org/wiki/Asymmetric_Laplace_distribution) as an `rv_continuous`.

### CODE REVIEW

File: scipy/stats/_continuous_distns.py
Comment (ev-br) [inline scipy/stats/_continuous_distns.py:4018]: Can use the `:doi:` sphinx role.

File: scipy/stats/_continuous_distns.py
Comment (ev-br) [inline scipy/stats/_continuous_distns.py:4024]: self._random_state can be an instance of either RandomState or Generator from rumpy.random.

**Review by ev-br — COMMENTED** [review]
For testing and doc generation, it needs a entry in https://github.com/scipy/scipy/blob/master/scipy/stats/_distr_params.py

File: scipy/stats/_continuous_distns.py
Comment (essandess) [reply scipy/stats/_continuous_distns.py]: Thanks done.

File: scipy/stats/_continuous_distns.py
Comment (essandess) [reply scipy/stats/_continuous_distns.py]: Thanks. Would you please check (and correct with code if necessary) my educated guess on the right way to handle this?

https://github.com/scipy/scipy/blob/899ed457f538bc1acf4126c85d18a22de8663e94/scipy/stats/_continuous_distns.py#L4024-L4025

**Comment by essandess** [comment]
> For testing and doc generation, it needs a entry in https://github.com/scipy/scipy/blob/master/scipy/stats/_distr_params.py

Thanks done.

**Comment by pvanmulbregt** [comment]
http://scipy.github.io/devdocs/dev/contributor/adding_new.html#adding-new describes all the steps for adding a new stats distribution.

File: scipy/stats/_continuous_distns.py
Comment (essandess) [reply scipy/stats/_continuous_distns.py]: Have this working now, thanks.

**Comment by essandess** [comment]
> http://scipy.github.io/devdocs/dev/contributor/adding_new.html#adding-new describes all the steps for adding a new stats distribution.

Thanks done.

**Comment by essandess** [comment]
> > http://scipy.github.io/devdocs/dev/contributor/adding_new.html#adding-new describes all the steps for adding a new stats distribution.
> 
> Thanks done.

I believe I’ve done all the necessary steps here. I see a failed build of the docs, but do not see that this commit is related: https://circleci.com/api/v1.1/project/github/scipy/scipy/20946/output/106/0?file=true&allocation-id=5f357866fb43201f425a2f52-0-build%2F6A550F56

I’ll need help if there’s anything more to be done.

File: scipy/stats/tests/test_distributions.py
Comment (pvanmulbregt) [inline scipy/stats/tests/test_distributions.py:460]: This tests generate Asymmetric Laplace against Laplace.  Are there any asymmetric tests that would be useful to perform?

File: doc/source/tutorial/stats/continuous_laplace_asymmetric.rst
Comment (pvanmulbregt) [inline doc/source/tutorial/stats/continuous_laplace_asymmetric.rst:19]: Are there other values that can be listed?  doc/source/tutorial/stats/continuous.rst lists many about 15 functions or attributes.

File: doc/source/tutorial/stats/continuous_laplace_asymmetric.rst
Comment (pvanmulbregt) [inline doc/source/tutorial/stats/continuous_laplace_asymmetric.rst:20]: Can you add a References section here prior to the Implementation? (E.g. see doc/source/tutorial/stats/continuous_kstwobign.rst for an example)
Ideally some which are readily accessible.  The reference in the implementation is behind a paywall.

File: doc/source/tutorial/stats/continuous_laplace_asymmetric.rst
Comment (essandess) [reply doc/source/tutorial/stats/continuous_laplace_asymmetric.rst:17]: Thanks done.

File: scipy/stats/tests/test_distributions.py
Comment (essandess) [reply scipy/stats/tests/test_distributions.py:460]: None of the asymmetric Laplace distributions yields known equivalent. I'm simply copying what `gennorm` already does and believe that it's an adequate test.

File: doc/source/tutorial/stats/continuous_laplace_asymmetric.rst
Comment (essandess) [reply doc/source/tutorial/stats/continuous_laplace_asymmetric.rst:19]: There are, but it gets complicated, and these are well-documented online. See e.g. https://en.wikipedia.org/wiki/Asymmetric_Laplace_distribution

File: doc/source/tutorial/stats/continuous_laplace_asymmetric.rst
Comment (essandess) [reply doc/source/tutorial/stats/continuous_laplace_asymmetric.rst:20]: Thanks done.

File: scipy/stats/tests/test_distributions.py
Comment (pvanmulbregt) [reply scipy/stats/tests/test_distributions.py:460]: Without a test of asymmetric Laplace , one could replace `stats.laplace_asymmetric` with `stats.laplace` and all tests would pass.  `stats.laplace_asymmetric` should agree with `stats.laplace` at `kappa=1`, but there must be some value of `kappa` where the `pdf/cdf/...` can be computed and compared against some stored values.

File: scipy/stats/_continuous_distns.py
Comment (pvanmulbregt) [inline scipy/stats/_continuous_distns.py:4024]: No need to abbreviate `kappa` to `kap` here.  `kappa` is the name of parameter in the docstring and the description of the distribution, so it makes sense to keep it as `kappa` in the code.

File: scipy/stats/tests/test_distributions.py
Comment (essandess) [reply scipy/stats/tests/test_distributions.py]: Done thanks.

File: scipy/stats/_continuous_distns.py
Comment (essandess) [reply scipy/stats/_continuous_distns.py]: Done thanks.

File: doc/source/tutorial/stats/continuous_laplace_asymmetric.rst
Comment (mdhaber) [inline doc/source/tutorial/stats/continuous_laplace_asymmetric.rst:2]:
```suggestion
.. _continuous-laplace-asymmetric:
```
to resolve doc build failure:
```
/home/circleci/repo/doc/source/tutorial/stats/continuous_laplace_asymmetric.rst:5: WARNING: duplicate label continuous-laplace, other instance in /home/circleci/repo/doc/source/tutorial/stats/continuous_laplace.rst
```

File: doc/source/tutorial/stats/continuous_laplace_asymmetric.rst
Comment (mdhaber) [inline doc/source/tutorial/stats/continuous_laplace_asymmetric.rst:33]: This is not rendering correctly:
image
Maybe you need to start the docstring with `r"""`?

File: scipy/stats/_continuous_distns.py
Comment (mdhaber) [inline scipy/stats/_continuous_distns.py:4115]: I checked and this does look equivalent to running the uniform random numbers through the `_ppf`. In that case, 

- If we can handle the overhead, why not just use the default rvs implementation?
- If we can't handle the overhead, why are the implementations of these functions different? Is there a property of uniformly distributed random numbers that makes this implementation tend to be more efficient than the implementation in `_ppf`?

**Review by mdhaber — CHANGES_REQUESTED** [review]
Looks quite close. Mostly just questions for the author and other reviewers and a few minor corrections.

File: doc/source/tutorial/stats/continuous_laplace_asymmetric.rst
Comment (mdhaber) [reply doc/source/tutorial/stats/continuous_laplace_asymmetric.rst:37]: @pvanmulbregt Is this satisfactory? I added a suggestion to fix `\sigma^2` -to `\mu_{2}`,  `\tilde\mu_3` to `\gamma_{1}`, and I added `\gamma_{2}`.

File: scipy/stats/tests/test_distributions.py
Comment (mdhaber) [reply scipy/stats/tests/test_distributions.py]: @essandess please let @pvanmulbregt or another reviewer decide whether to click the resolve button.  It helps us keep track of where the review is and lets us check whether the changes are sufficient.

I unresolved this because I'm not sure that the comment has been addressed.  @pvanmulbregt, has this been resolved? I don't think I can see the commit @pvanmulbregt reviewed because of the force pushes, but I think `test_asymmetric_laplace` is new, and it does test an identity of the pdf. However, @pvanmulbregt seemed to want to compare againsts reference value, e.g. computed by R or published online.

**Comment by mdhaber** [comment]
I also wanted to mention that I checked this against the [guide](http://scipy.github.io/devdocs/dev/contributor/adding_new.html#adding-new) for adding new distributions, and it seems to satisfy the checklist. Step 4 and 5 are satisfied because the distribution does have infinite support, and step 6 is satisfied because the only requirement of the shape parameter is that it is positive. 7 is ok because `squirrel_gen.ppf()` is not expensive to computer. Re: 8, `squirrel_gen.rvs()` is not expensive given `_ppf`, so I've suggested that we might want to just rely on the default implementation or at least unify the `_rvs` and `_ppf` code. Tests, a tutorial, and sane parameters have been added in appropriate places, etc.

File: doc/source/tutorial/stats/continuous_laplace_asymmetric.rst
Comment (essandess) [reply doc/source/tutorial/stats/continuous_laplace_asymmetric.rst:33]: I'm not what you mean—this is an `.rst` file, not ` Python script. I replaced this with the UTF-8 ó and hope that it works.

File: doc/source/tutorial/stats/continuous_laplace_asymmetric.rst
Comment (essandess) [reply doc/source/tutorial/stats/continuous_laplace_asymmetric.rst:22]: Whoops thanks. I checked to make sure that the sign is correct in the actual code; it is.

File: doc/source/tutorial/stats/continuous_laplace_asymmetric.rst
Comment (essandess) [reply doc/source/tutorial/stats/continuous_laplace_asymmetric.rst:4]: Done thanks.

File: scipy/stats/_continuous_distns.py
Comment (WarrenWeckesser) [inline scipy/stats/_continuous_distns.py:4107]: Use Unicode here:
```suggestion
    .. [1] Kozubowski TJ and Podgórski K. A Multivariate and
```
We have a restricted subset of Unicode characters that we allow in the docstrings, and currently ó is not in that set. It is a widely used latin-1 character, though, so we can add it.  This means you'll have to add the character to the `allowed` set here:

https://github.com/scipy/scipy/blob/06f51e32837b4f78f2630701091cd98ff4c6ddf7/tools/unicode-check.py#L13

File: doc/source/tutorial/stats/continuous_laplace_asymmetric.rst
Comment (WarrenWeckesser) [reply doc/source/tutorial/stats/continuous_laplace_asymmetric.rst:33]: This file still shows `Podg\'orski`.  Did you push the change in a commit?

File: scipy/stats/_continuous_distns.py
Comment (WarrenWeckesser) [inline scipy/stats/_continuous_distns.py:4143]: PEP-8 [indentation](https://www.python.org/dev/peps/pep-0008/#indentation) and whitespace:
```suggestion
        return np.where(x >= 0,
                        np.exp(-x*kappa)*(kapinv/kappkapinv),
                        1 - np.exp(x*kapinv)*(kappa/kappkapinv))
```

File: scipy/stats/_continuous_distns.py
Comment (WarrenWeckesser) [inline scipy/stats/_continuous_distns.py:4150]: PEP-8 [indentation](https://www.python.org/dev/peps/pep-0008/#indentation) and whitespace:
```suggestion
        return np.where(q >= kappa/kappkapinv,
                        -np.log((1 - q)*kappkapinv*kappa)*kapinv,
                        np.log(q*kappkapinv/kappa)*kappa)
```

**Review by WarrenWeckesser — CHANGES_REQUESTED** [review]
Overall this looks pretty good.  I suggested some changes inline.  There should also be some simple, explicit tests for the `pdf`, `cdf`, `ppf`, `sf` and `isf` methods using known values. For example, a quick calculation "by hand" gives us

    x        kappa    PDF       CDF
    -------- -----   -----   ---------
    -log(16)   2     1/10      1/5
    log(10)    2     1/250   1 - 1/500

Those values can be used to create some simple tests for the `pdf`, `cdf`, `ppf` `sf` and `isf` methods that use fairly small relative tolerances (e.g. 1e-13 or so).

File: scipy/stats/_continuous_distns.py
Comment (essandess) [reply scipy/stats/_continuous_distns.py:4106]: Done thanks.

File: scipy/stats/_continuous_distns.py
Comment (essandess) [reply scipy/stats/_continuous_distns.py:4138]: The correct answer from the reviewers will (should) be that the current implementation is preferred because it avoids roundoff errors: compare `exp(-x)` to `1-(1-exp(-x))` for large `x`.

File: scipy/stats/_continuous_distns.py
Comment (essandess) [reply scipy/stats/_continuous_distns.py:4145]: This is a trivial derivation; no reference is necessary.

File: doc/source/tutorial/stats/continuous_laplace_asymmetric.rst
Comment (essandess) [reply doc/source/tutorial/stats/continuous_laplace_asymmetric.rst:23]: Notation changed and excess kurtosis added thanks.

File: scipy/stats/_continuous_distns.py
Comment (essandess) [reply scipy/stats/_continuous_distns.py:4115]: I'm not familiar enough with `stats` internals to comment. and defer to you all.

File: scipy/stats/_continuous_distns.py
Comment (essandess) [reply scipy/stats/_continuous_distns.py:4107]: Done thanks.

File: scipy/stats/_continuous_distns.py
Comment (essandess) [reply scipy/stats/_continuous_distns.py:4127]: Re: `_rvs`, does this simply mean deleting this redefined method from the class? How  then is  the shape parameter handled (out of curiosity?)

File: scipy/stats/_continuous_distns.py
Comment (mdhaber) [reply scipy/stats/_continuous_distns.py:4138]: Gotcha. Order of operations.

File: doc/source/tutorial/stats/continuous_laplace_asymmetric.rst
Comment (mdhaber) [reply doc/source/tutorial/stats/continuous_laplace_asymmetric.rst:4]: I don't see any changes to the text in the tutorial. Please check that you pushed the changes.

File: doc/source/tutorial/stats/continuous_laplace_asymmetric.rst
Comment (mdhaber) [reply doc/source/tutorial/stats/continuous_laplace_asymmetric.rst:23]: Similarly here; looks like a push is needed. Please avoid the force pushes until the very end (e.g. if you want to clean up commit history or rebase on master).

File: scipy/stats/_continuous_distns.py
Comment (mdhaber) [reply scipy/stats/_continuous_distns.py:4127]: Yes. If you remove your `_rvs` override, then the default implementation will be used. Here's the default implementation (top comments removed):
```
    def _rvs(self, *args, size=None, random_state=None):
        ## Use basic inverse cdf algorithm for RV generation as default.
        U = random_state.uniform(size=size)
        Y = self._ppf(U, *args)
        return Y
```
It handles the shape by passing it into `_ppf`. I think the results will be almost identical because this is basically what your `_rvs` method already does (except the ppf code is rewritten a bit differently).

**Comment by mdhaber** [comment]
@essandess this is really close! If you push your changes that respond to these comments, we can probably get this in before 1.6.0 branches.

File: scipy/stats/_continuous_distns.py
Comment (essandess) [reply scipy/stats/_continuous_distns.py:4143]: Done thanks.

File: scipy/stats/_continuous_distns.py
Comment (essandess) [reply scipy/stats/_continuous_distns.py:4150]: Done thanks.

File: scipy/stats/_continuous_distns.py
Comment (essandess) [reply scipy/stats/_continuous_distns.py:4157]: Done thanks.

File: doc/source/tutorial/stats/continuous_laplace_asymmetric.rst
Comment (essandess) [reply doc/source/tutorial/stats/continuous_laplace_asymmetric.rst]: Just pushed all my commits.

File: scipy/stats/_continuous_distns.py
Comment (essandess) [reply scipy/stats/_continuous_distns.py]: This mod to `_rvs` deleted per discussion below.

**Comment by mdhaber** [comment]
@essandess The force push has erased the history, so I can't see what has changed since I last reviewed. I will check the unresolved comments, but someone else will have to merge.

File: scipy/stats/tests/test_distributions.py
Comment (mdhaber) [inline scipy/stats/tests/test_distributions.py:542]:
```suggestion
```

File: scipy/stats/tests/test_distributions.py
Comment (mdhaber) [inline scipy/stats/tests/test_distributions.py:527]:
```suggestion
        assert_allclose(pdf1, pdf2)
```
[NumPy suggests](https://numpy.org/doc/stable/reference/generated/numpy.testing.assert_almost_equal.html) transitioning to `assert_allclose`.

File: scipy/stats/tests/test_distributions.py
Comment (mdhaber) [inline scipy/stats/tests/test_distributions.py:536]:
```suggestion
        assert_allclose(pdf1, pdf2)
```

File: scipy/stats/tests/test_distributions.py
Comment (mdhaber) [inline scipy/stats/tests/test_distributions.py:547]:
```suggestion
        assert_allclose(np.concatenate((pdf1, cdf1)), np.concatenate((pdf1, cdf1)))
```

File: scipy/stats/tests/test_distributions.py
Comment (mdhaber) [inline scipy/stats/tests/test_distributions.py:538]: These new tests look good.

**Comment by mdhaber** [comment]
Doctest is failing:
```
File "doc/source/tutorial/stats.rst", line 100, in /home/vsts/work/1/s/tools/../doc/source/tutorial/stats.rst
Failed example:
    print('number of continuous distributions: %d' % len(dist_continu))
Expected:
    number of continuous distributions: 100
Got:
    number of continuous distributions: 101
```
I thought this was already fixed. Did this revert?

File: doc/source/tutorial/stats/continuous_laplace_asymmetric.rst
Comment (mdhaber) [inline doc/source/tutorial/stats/continuous_laplace_asymmetric.rst:30]:
```suggestion
    \gamma_1 & = & \frac{2(1-\kappa^6)}{(1+\kappa^4)^{3/2}}\\
```

Comment (mdhaber) [reply doc/source/tutorial/stats/continuous_laplace_asymmetric.rst:30]: gamma_2 is on the same line

File: scipy/stats/tests/test_distributions.py
Comment (essandess) [reply scipy/stats/tests/test_distributions.py:542]: Done thanks.

File: scipy/stats/tests/test_distributions.py
Comment (essandess) [reply scipy/stats/tests/test_distributions.py:527]: Done thanks.

File: scipy/stats/tests/test_distributions.py
Comment (essandess) [reply scipy/stats/tests/test_distributions.py:536]: Done thanks.

File: scipy/stats/tests/test_distributions.py
Comment (essandess) [reply scipy/stats/tests/test_distributions.py:547]: Done thanks.

File: scipy/stats/tests/test_distributions.py
Comment (essandess) [reply scipy/stats/tests/test_distributions.py:538]: Added thanks.

File: doc/source/tutorial/stats/continuous_laplace_asymmetric.rst
Comment (essandess) [reply doc/source/tutorial/stats/continuous_laplace_asymmetric.rst:30]: Done thanks.

**Review by mdhaber — APPROVED** [review]
It looks like all of my comments have been addressed. @pvanmulbregt would you be willing to take a second pass? I think your comments were addressed, too.

File: scipy/stats/_continuous_distns.py
Comment (WarrenWeckesser) [inline scipy/stats/_continuous_distns.py:4110]: Use the Unicode character here, too.  (This is the change that also requires the change in `unicode-check.py`.)
```suggestion
    .. [2] Kozubowski TJ and Podgórski K. A Multivariate and
```

**Review by WarrenWeckesser — CHANGES_REQUESTED** [review]
I still have one tiny change request that was missed in the last update.  I'd push it myself, but I generally only do that when I'm going to also merge the PR.

I think the updated tests meet @pvanmulbregt's request for tests that are specific to the asymmetric laplace distribution, but I'd like to get a thumb's up from @pvanmulbregt before merging.

File: scipy/stats/_continuous_distns.py
Comment (essandess) [reply scipy/stats/_continuous_distns.py]: Whoops, added now, thanks.

**Review by WarrenWeckesser — APPROVED** [review]
Thanks @essandess.  I'll give this another day and merge if there are no more comments.

**Comment by WarrenWeckesser** [comment]
Merged, thanks @essandess!

I added `laplace_asymmetric` to the list of new distributions in the [release notes](https://github.com/scipy/scipy/wiki/Release-note-entries-for-SciPy-1.6.0).

**Pull request merged**
